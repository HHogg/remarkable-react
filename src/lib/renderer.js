const React = require('react');
const buildTokenTree = require('./token-tree-builder');

const defaultOptions = {
  keyGen: (token, index) => index,
};

const defaultComponents = {
  a: 'a',
  blockquote: 'blockquote',
  br: 'br',
  code: 'code',
  del: 'del',
  em: 'em',
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  hr: 'hr',
  img: 'img',
  ins: 'ins',
  li: 'li',
  mark: 'mark',
  ol: 'ol',
  p: 'p',
  pre: 'pre',
  strong: 'strong',
  sub: 'sub',
  sup: 'sup',
  table: 'table',
  tbody: 'tbody',
  td: 'td',
  th: 'th',
  thead: 'thead',
  tr: 'tr',
  ul: 'ul',
};

const defaultRemarkableProps = {
  align: (align) => ({
    key: 'style',
    value: align && { textAlign: align },
  }),
  alt: true,
  block: false,
  content: false,
  hLevel: false,
  href: true,
  level: false,
  lines: false,
  order: false,
  params: false,
  src: true,
  tight: false,
  title: true,
  type: false,
};

const defaultTokens = {
  blockquote_open: 'blockquote',
  bullet_list_open: 'ul',
  code: 'code',
  del_open: 'del',
  em_open: 'em',
  fence: ['pre', 'code'],
  hardbreak: 'br',
  heading_open: (token) => `h${token.hLevel}`,
  hr: 'hr',
  image: 'img',
  ins_open: 'ins',
  link_open: 'a',
  list_item_open: 'li',
  mark_open: 'mark',
  ordered_list_open: 'ol',
  paragraph_open: 'p',
  softbreak: 'br',
  strong_open: 'strong',
  sub: 'sub',
  sup: 'sup',
  table_open: 'table',
  tbody_open: 'tbody',
  td_open: 'td',
  th_open: 'th',
  thead_open: 'thead',
  tr_open: 'tr',
};

function Renderer(options = {}) {
  this.options = {
    ...defaultOptions,
    ...options,

    components: {
      ...defaultComponents,
      ...(options.components || {}),
    },

    remarkableProps: {
      ...defaultRemarkableProps,
      ...(options.remarkableProps || {}),
    },

    tokens: {
      ...defaultTokens,
      ...(options.tokens || {}),
    }
  };
}

Renderer.prototype.render = function(tokens = [], remarkableOptions) {
  return this.renderTokenTree(
    buildTokenTree(this.options.tokens, tokens),
    remarkableOptions);
}

Renderer.prototype.renderTokenTree = function(tokens, remarkableOptions) {
  if (!tokens || !Array.isArray(tokens)) {
    return tokens;
  }

  return tokens.map((token, index) => {
    return this.options.components[token.type]
      ? React.createElement(
          this.options.components[token.type],
          this.getTokenProps(token, index, remarkableOptions),
          this.renderTokenTree(token.children, remarkableOptions))
      : token.children;
  });
}

Renderer.prototype.getTokenProps = function(token, index, remarkableOptions) {
  const { props, type } = token;
  const tokenProps = {
    key: this.options.keyGen(token, index),
  };

  if (typeof this.options.components[type] === 'function') {
    tokenProps.options = remarkableOptions;
  }

  Object.keys(props).forEach((prop) => {
    if (this.options.remarkableProps.hasOwnProperty(prop)) {
      if (typeof this.options.remarkableProps[prop] === 'function') {
        const { key = prop, value = props[prop] } = this.options.remarkableProps[prop](props[prop]);

        if (value) {
          tokenProps[key] = value;
        }
      } else if (typeof this.options.remarkableProps[prop] === 'string') {
        tokenProps[this.options.remarkableProps[prop]] = props[prop];
      } else if (this.options.remarkableProps[prop] === true ||
          typeof this.options.components[type] === 'function') {
        tokenProps[prop] = props[prop];
      }
    }
  });

  return tokenProps;
}

module.exports = Renderer;


