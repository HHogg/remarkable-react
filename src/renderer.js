import React from 'react';
import buildTokenTree from './token-tree-builder';
import { assign } from './utils';

const defaultOptions = {
  keyGen: (token, index) => index,
};

const defaultComponents = {
  a: 'a',
  blockquote: 'blockquote',
  br: 'br',
  code: 'code',
  del: 'del',
  em: 'em',
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  hr: 'hr',
  img: 'img',
  ins: 'ins',
  li: 'li',
  mark: 'mark',
  ol: 'ol',
  p: 'p',
  pre: 'pre',
  strong: 'strong',
  sub: 'sub',
  sup: 'sup',
  table: 'table',
  tbody: 'tbody',
  td: 'td',
  th: 'th',
  thead: 'thead',
  tr: 'tr',
  ul: 'ul',
};

const defaultRemarkableProps = {
  align: (align) => ({
    key: 'style',
    value: align && { textAlign: align },
  }),
  alt: true,
  block: false,
  content: false,
  hLevel: false,
  href: true,
  level: false,
  lines: false,
  linkTarget: (target, type) => type === 'a' && ({ key: 'target' }),
  order: false,
  params: false,
  src: true,
  tight: false,
  title: true,
  type: false,
};

const defaultTokens = {
  blockquote_open: 'blockquote',
  bullet_list_open: 'ul',
  code: 'code',
  del_open: 'del',
  em_open: 'em',
  fence: ['pre', 'code'],
  hardbreak: 'br',
  heading_open: (token) => `h${token.hLevel}`,
  hr: 'hr',
  image: 'img',
  ins_open: 'ins',
  link_open: 'a',
  list_item_open: 'li',
  mark_open: 'mark',
  ordered_list_open: 'ol',
  paragraph_open: 'p',
  softbreak: (_, options) => options.breaks ? 'br' : undefined,
  strong_open: 'strong',
  sub: 'sub',
  sup: 'sup',
  table_open: 'table',
  tbody_open: 'tbody',
  td_open: 'td',
  th_open: 'th',
  thead_open: 'thead',
  tr_open: 'tr',
};

export default class Renderer {
  constructor(options = {}) {
    this.options = assign({}, defaultOptions, options, {
      components: assign({}, defaultComponents, options.components),
      remarkableProps: assign({}, defaultRemarkableProps, options.remarkableProps),
      tokens: assign({}, defaultTokens, options.tokens),
    });
  }

  render(tokens = [], remarkableOptions) {
    return this.renderTokenTree(
      buildTokenTree(this.options.tokens, tokens, remarkableOptions),
      remarkableOptions);
  }

  renderTokenTree(tokens, remarkableOptions) {
    if (!tokens || !Array.isArray(tokens)) {
      return tokens;
    }

    return tokens.map((token, index) => {
      return this.options.components[token.type]
        ? React.createElement(
            this.options.components[token.type],
            this.getTokenProps(token, index, remarkableOptions),
            this.renderTokenTree(token.children, remarkableOptions))
        : token.children;
    });
  }

  getTokenProp(type, prop, propValue) {
    if (this.options.remarkableProps.hasOwnProperty(prop)) {
      if (typeof this.options.remarkableProps[prop] === 'function') {
        const keyValue = this.options.remarkableProps[prop](propValue, type);

        if (keyValue) {
          return {
            key: keyValue.key || prop,
            value: keyValue.value || propValue,
          };
        }
      } else if (typeof this.options.remarkableProps[prop] === 'string') {
        return {
          key: this.options.remarkableProps[prop],
          value: propValue,
        }
      } else if (this.options.remarkableProps[prop] === true ||
          typeof this.options.components[type] === 'function') {
        return {
          key: prop,
          value: propValue,
        }
      }
    }
  }

  getTokenProps(token, index, remarkableOptions) {
    const { props, type } = token;
    const tokenProps = {
      key: this.options.keyGen(token, index),
    };

    if (typeof this.options.components[type] === 'function') {
      tokenProps.options = remarkableOptions;
    }

    [props, remarkableOptions].forEach((props) => {
      Object.keys(props).forEach((prop) => {
        const tokenProp = this.getTokenProp(type, prop, props[prop]);
        if (tokenProp) tokenProps[tokenProp.key] = tokenProp.value;
      });
    });

    return tokenProps;
  }
}
