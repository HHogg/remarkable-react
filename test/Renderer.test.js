import renderer from 'react-test-renderer';
import buildReactTree from './buildReactTree';

const renderToJson = (...args) =>
   renderer.create(buildReactTree(...args)).toJSON();

describe('Renderer components', () => {
  test('<a>', () => {
    expect(renderToJson(`

[Link](http://link.com "Title")

    `)).toMatchSnapshot();
  });

  test('<blockquote>', () => {
    expect(renderToJson(`

> Blockquote

    `)).toMatchSnapshot();
  });

  describe('<br>', () => {
    test('with breaks', () => {
      expect(renderToJson(`

New
Line

      `, { breaks: true })).toMatchSnapshot();
    });

    test('without breaks', () => {
      expect(renderToJson(`

New
Line

      `)).toMatchSnapshot();
    });
  });

  test('<code>', () => {
    expect(renderToJson(`

\`Code\`

    `)).toMatchSnapshot();
  });

  test('<del>', () => {
    expect(renderToJson(`

~~Deleted~~

    `)).toMatchSnapshot();
  });

  test('<em>', () => {
    expect(renderToJson(`

*Italic*

    `)).toMatchSnapshot();
  });

  describe('<footnote>', () => {
    test('inline', () => {
      expect(renderToJson(`

Footnote^[Foonote body].

      `)).toMatchSnapshot();
    });

    test('separated', () => {
      expect(renderToJson(`

Footnote[^first].

[^first]: #### Footnote

    Footnote body

      `)).toMatchSnapshot();
    });
  });

  test('<h1>', () => {
    expect(renderToJson(`

# Heading 1

    `)).toMatchSnapshot();
  });

  test('<h2>', () => {
    expect(renderToJson(`

## Heading 2

    `)).toMatchSnapshot();
  });

  test('<h3>', () => {
    expect(renderToJson(`

### Heading 3

    `)).toMatchSnapshot();
  });

  test('<h4>', () => {
    expect(renderToJson(`

#### Heading 4

    `)).toMatchSnapshot();
  });

  test('<h5>', () => {
    expect(renderToJson(`

##### Heading 5

    `)).toMatchSnapshot();
  });

  test('<h6>', () => {
    expect(renderToJson(`

###### Heading 6

    `)).toMatchSnapshot();
  });

  test('<hr>', () => {
    expect(renderToJson(`

___

    `)).toMatchSnapshot();
  });

  describe('<html>', () => {
    test('with html', () => {
      expect(renderToJson(`

<p>Hello</p>
<iframe src="/something.html" />

      `, { html: true })).toMatchSnapshot();
    });

    test('without html', () => {
      expect(renderToJson(`

<p>Hello</p>
<iframe src="/something.html" />

      `)).toMatchSnapshot();
    });
  });

  test('<img>', () => {
    expect(renderToJson(`

![Alt](http://image.com "Title")

    `)).toMatchSnapshot();
  });

  test('<ins>', () => {
    expect(renderToJson(`

++Insert++

    `)).toMatchSnapshot();
  });

  test('<mark>', () => {
    expect(renderToJson(`

==Mark==

    `)).toMatchSnapshot();
  });

  test('<ol> <li>', () => {
    expect(renderToJson(`

1. Level 1 Item 1
2. Level 1 Item 2
3. Level 1 Item 3

    `)).toMatchSnapshot();
  });

  test('<p>', () => {
    expect(renderToJson(`

Paragraph

    `)).toMatchSnapshot();
  });

  test('<pre> <code>', () => {
    expect(renderToJson(`

\`\`\`
Code
\`\`\`

    `)).toMatchSnapshot();
  });

  test('<strong>', () => {
    expect(renderToJson(`

**Strong**

    `)).toMatchSnapshot();
  });

  test('<sub>', () => {
    expect(renderToJson(`

~Subscript~

    `)).toMatchSnapshot();
  });

  test('<sup>', () => {
    expect(renderToJson(`

^Superscript^

    `)).toMatchSnapshot();
  });

  test('<table> <thead> <tbody> <tr> <th> <td>', () => {
    expect(renderToJson(`

| Header 1 | Header 2 |
| -------- | -------- |
| Row 1 Cell 1 | Row 1 Cell 2 |
| Row 2 Cell 1 | Row 2 Cell 2 |

    `)).toMatchSnapshot();
  });

  test('<ul> <li>', () => {
    expect(renderToJson(`

* Level 1 Item 1
* Level 1 Item 2
  * Level 2 Item 1
    * Level 3 Item 1
  * Level 2 Item 2

    `)).toMatchSnapshot();
  });

  test('CustomComponent', () => {

  });
});
