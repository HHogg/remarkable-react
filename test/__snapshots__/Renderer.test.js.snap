// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer components <a> 1`] = `
<p>
  <a
    href="http://link.com"
    title="Title"
  >
    Link
  </a>
</p>
`;

exports[`Renderer components <blockquote> 1`] = `
<blockquote>
  <p>
    Blockquote
  </p>
</blockquote>
`;

exports[`Renderer components <br> with breaks 1`] = `
<p>
  New
  <br />
  Line
</p>
`;

exports[`Renderer components <br> without breaks 1`] = `
<p>
  New
  Line
</p>
`;

exports[`Renderer components <code> 1`] = `
<p>
  <code>
    Code
  </code>
</p>
`;

exports[`Renderer components <del> 1`] = `
<p>
  <del>
    Deleted
  </del>
</p>
`;

exports[`Renderer components <em> 1`] = `
<p>
  <em>
    Italic
  </em>
</p>
`;

exports[`Renderer components <footnote> inline 1`] = `
Array [
  <p>
    Footnote
    <sup
        id="fn0:0"
    >
        <a
            href="#fn0"
        >
            [1]
        </a>
    </sup>
    .
</p>,
  <section>
    <ol>
        <li
            id="fn0"
        >
            <p>
                Foonote body
                <a
                    href="#fn0:0"
                >
                    ↩
                </a>
            </p>
        </li>
    </ol>
</section>,
]
`;

exports[`Renderer components <footnote> separated 1`] = `
Array [
  <p>
    Footnote
    <sup
        id="fn0:0"
    >
        <a
            href="#fn0"
        >
            [1]
        </a>
    </sup>
    .
</p>,
  <section>
    <ol>
        <li
            id="#fn0"
        >
            <h4>
                Footnote
            </h4>
            <p>
                Footnote body
                <a
                    href="fn0:0"
                >
                    ↩
                </a>
            </p>
        </li>
    </ol>
</section>,
]
`;

exports[`Renderer components <h1> 1`] = `
<h1>
  Heading 1
</h1>
`;

exports[`Renderer components <h2> 1`] = `
<h2>
  Heading 2
</h2>
`;

exports[`Renderer components <h3> 1`] = `
<h3>
  Heading 3
</h3>
`;

exports[`Renderer components <h4> 1`] = `
<h4>
  Heading 4
</h4>
`;

exports[`Renderer components <h5> 1`] = `
<h5>
  Heading 5
</h5>
`;

exports[`Renderer components <h6> 1`] = `
<h6>
  Heading 6
</h6>
`;

exports[`Renderer components <hr> 1`] = `<hr />`;

exports[`Renderer components <html> with html 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p>Hello</p>
    <iframe src=\\"/something.html\\" />
    ",
    }
  }
/>
`;

exports[`Renderer components <html> without html 1`] = `
<p>
  &lt;p&gt;Hello&lt;/p&gt;
  &lt;iframe src="/something.html" /&gt;
</p>
`;

exports[`Renderer components <img> 1`] = `
<p>
  <img
    alt="Alt"
    src="http://image.com"
    title="Title"
  />
</p>
`;

exports[`Renderer components <ins> 1`] = `
<p>
  <ins>
    Insert
  </ins>
</p>
`;

exports[`Renderer components <mark> 1`] = `
<p>
  <mark>
    Mark
  </mark>
</p>
`;

exports[`Renderer components <ol> <li> 1`] = `
<ol>
  <li>
    <p>
      Level 1 Item 1
    </p>
  </li>
  <li>
    <p>
      Level 1 Item 2
    </p>
  </li>
  <li>
    <p>
      Level 1 Item 3
    </p>
  </li>
</ol>
`;

exports[`Renderer components <p> 1`] = `
<p>
  Paragraph
</p>
`;

exports[`Renderer components <pre> <code> 1`] = `
<pre>
  <code>
    Code
    
  </code>
</pre>
`;

exports[`Renderer components <strong> 1`] = `
<p>
  <strong>
    Strong
  </strong>
</p>
`;

exports[`Renderer components <sub> 1`] = `
<p>
  <sub>
    Subscript
  </sub>
</p>
`;

exports[`Renderer components <sup> 1`] = `
<p>
  <sup>
    Superscript
  </sup>
</p>
`;

exports[`Renderer components <table> <thead> <tbody> <tr> <th> <td> 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Header 1
      </th>
      <th>
        Header 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Row 1 Cell 1
      </td>
      <td>
        Row 1 Cell 2
      </td>
    </tr>
    <tr>
      <td>
        Row 2 Cell 1
      </td>
      <td>
        Row 2 Cell 2
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Renderer components <ul> <li> 1`] = `
<ul>
  <li>
    <p>
      Level 1 Item 1
    </p>
  </li>
  <li>
    <p>
      Level 1 Item 2
    </p>
    <ul>
      <li>
        <p>
          Level 2 Item 1
        </p>
        <ul>
          <li>
            <p>
              Level 3 Item 1
            </p>
          </li>
        </ul>
      </li>
      <li>
        <p>
          Level 2 Item 2
        </p>
      </li>
    </ul>
  </li>
</ul>
`;
